apps/managedidentity/managedidentity.go

```diff
@@ -173,17 +280,104 @@ func createIMDSAuthRequest(ctx context.Context, id ID, resource string, claims s
 }
 
-func (client Client) getTokenForRequest(req *http.Request) (accesstokens.TokenResponse, error) {
+func createAzureArcAuthRequest(ctx context.Context, id ID, resource string, claims string) (*http.Request, error) {
+	identityEndpoint, _ := getAzureArcEnvironmentVariables()
+	var msiEndpoint *url.URL
+
+	if identityEndpoint == "" {
+		identityEndpoint = azureArcEndpoint
+	}
+
+	if _, ok := id.(systemAssignedValue); !ok {
+		return nil, errors.New("Azure Arc doesn't support specifying a user-assigned managed identity at runtime")
+	}
+
+	identityEndpoint = strings.Replace(identityEndpoint, "localhost", "127.0.0.1", -1)
+	msiEndpoint, parseErr := url.Parse(identityEndpoint)
+	if parseErr != nil {
+		return nil, fmt.Errorf("couldn't parse %q: %s", identityEndpoint, parseErr)
+	}
+
+	msiParameters := msiEndpoint.Query()
+	msiParameters.Set(apiVersionQueryParameterName, azureArcAPIVersion)
+	resource = strings.TrimSuffix(resource, "/.default")
+	msiParameters.Set(resourceQueryParameterName, resource)
+
+	if len(claims) > 0 {
+		msiParameters.Set("claims", claims)
+	}
+
+	msiEndpoint.RawQuery = msiParameters.Encode()
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, msiEndpoint.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating http request %s", err)
+	}
+	req.Header.Set(metaHTTPHeaderName, "true")
+	return req, nil
+}
+
+// GetEnvironmentVariables returns the identity and IMDS endpoints
+func getAzureArcEnvironmentVariables() (string, string) {
+	identityEndpoint := os.Getenv(IdentityEndpointEnvVar)
+	imdsEndpoint := os.Getenv(ArcIMDSEnvVar)
+
+	if identityEndpoint == "" || imdsEndpoint == "" {
+		platform := runtime.GOOS
+		fileDetectionPath, retrievedPath := azureArcFileDetection[platform]
+
+		if retrievedPath {
+			if _, err := os.Stat(fileDetectionPath); err == nil {
+				identityEndpoint = azureArcEndpoint
+				imdsEndpoint = himdsExecutableHelperString
+			}
+		}
+	}
+	return identityEndpoint, imdsEndpoint
+}
+
+func validateAzureArcEnvironment(identityEndpoint, imdsEndpoint string, platform string) bool {
+	if identityEndpoint != "" && imdsEndpoint != "" {
+		return true
+	}
+
+	himdsFilePath, platformSupported := azureArcFileDetection[platform]
+
+	if platformSupported && fileExists(himdsFilePath) {
+		return true
+	}
+
+	return false
+}
+
+func fileExists(path string) bool {
+	if _, err := os.Stat(path); err != nil {
+		if os.IsNotExist(err) {
+			return false
+		}
+	}
+	return true
+}
+
+func (client Client) getTokenForRequest(ctx context.Context, req *http.Request, resHandler responseHandler) (accesstokens.TokenResponse, error) {
+	var r accesstokens.TokenResponse
+
 	resp, err := client.httpClient.Do(req)
 	if err != nil {
 		return accesstokens.TokenResponse{}, err
 	}
+
 	responseBytes, err := io.ReadAll(resp.Body)
 	defer resp.Body.Close()
+
 	if err != nil {
 		return accesstokens.TokenResponse{}, err
 	}
+
 	switch resp.StatusCode {
+	case http.StatusUnauthorized:
+		if resHandler != nil {
```

Comment:
I suggest sending the key discovery request somewhere else and leaving `getTokenForRequest` entirely responsible for token requests. The complexity here doesn't seem worthwhile to me:
- "response handler" implies it applies to all responses but really it's an "authentication challenge handler"
- this looks like a general purpose mechanism but it applies only to Arc
- the callback for Arc recurses into this function; what happens when the server responds with another 401?